probe: add backend for logging via probes

kv2019i opened the pull request:
probe logging: add backend for logging via probes
    
    Work in progress code that routes Zephyr logs via the logging
    subsystem to SOF probe extraction point.
    
    Lot of hardcodes still in this version, but this is tested to work
    with Linux SOF driver, crecord and sof/tools/probes for decoding.
    
    Example usage in Linux:
    
    term1# crecord -c3 -d0 -b8192 -f4 -FS32_LE -R48000 -C4 /tmp/extract.dat
    term2# echo 0,3,0 > /sys/kernel/debug/sof/probe_points
    term1# <terminate crecord>
    term1# sof-probes -p /tmp/extract.dat
    term1# cat buffer_999999.wav

kv2019i commented:
FYI @teburd

lgirdwood commented:
@kv2019i we should be able to make this scriptable in sof-tools. We do need a separate demux tool though that can split pcm/log/trace data by slot into individual files.

kv2019i commented:
@lgirdwood wrote:

> @kv2019i we should be able to make this scriptable in sof-tools. We do need a separate demux tool though that can split pcm/log/trace data by slot into individual files.

sof/tools/probes/probes_main.c is that demux tool, so we have that already.

We might need a more integrated tool. One developer requirement is to observe "live" logs from the FW and the current palette of tools forces to cut the logging to see any decoded logs. So potentially we need a more unified tool that combines probe point configuration, reading data from compress PCM and calling log decoders (e.g. the zephyr/mipi tool for syst catalog). In the end, developer wants to see live logs in human-readable format on her console. FYI @marc-hb  @jsarha @pjdobrowolski

lgirdwood commented:
> We might need a more integrated tool. One developer requirement is to observe "live" logs from the FW and the current palette of tools forces to cut the logging to see any decoded logs. So potentially we need a more unified tool that combines probe point configuration, reading data from compress PCM and calling log decoders (e.g. the zephyr/mipi tool for syst catalog). In the end, developer wants to see live logs in human-readable format on her console. FYI @marc-hb @jsarha @pjdobrowolski

ok, we can update the existing tool to strip the container for slot N and send it's stdout to log/trace tool stdin as a cmd line option.

kv2019i commented:
V2 update: 
 - fixed a few glaring bugs
 - some cleanup to code
 - renamed special buffer-id to 999999 (99 could be used in fairly typical topologies)

kv2019i commented on src/probe/probe.c:
```diff
@@ -906,6 +906,8 @@ static void probe_cb_free(void *arg, enum notify_id type, void *data)
 }
 
+extern struct ipc_comp_dev logging_comp;
+
```
This is ugly and not sure if we want to keep this. I'll made one try to get rid of this, but it broke the notifier logic. I'll retry again. Alternative we could drop have more explicit callbacks between the backend and probe client code, and not rely on comp_buffer struct and the notifier.h interface.

kv2019i replied:
V3 removed this usage of ipc_comp_dev.

kv2019i commented on src/include/sof/probe/probe.h:
```diff
@@ -12,6 +12,16 @@
+ *     prototype, a magic buffer identifier that is known
+ *     both by host and the FW, is used.
+ */
+#define PROBE_LOGGING_BUFFER_ID_MAGIC     999999
```
If we decide to make a mergeable version for IPC3, perhaps the best approach would be extend the ".purpose" field of current probe point interface. I.e. you'd set up a probe point with a special purpose and this would be picked up FW to set up a probe point for logging. Any magic numbers could be hidden within the FW.

ujfalusi replied:
What about using `BIT(31)` in buffer_id as a flag? If it is set then the rest of the buffer_id is ignored and it is for logging.
If you will:
```
union {
	uin32_t buffer_id;
	struct {
		uin32_t id : 31; /* buffer id if logging is 0 */
		uint32_t logging : 1;
	} fields;
};
```
or add new purpose, like `PROBE_PURPOSE_LOGGING` on top of EXTRACTION and INJECTION?

kv2019i replied:
New purpose added in V3 patch.

kv2019i commented on src/trace/log_backend_probe.c:
```diff
@@ -0,0 +1,138 @@
+	return flags;
+}
+
+#undef panic
```
Ugh, SOF headers redefine panic(). Took a while to figure out why code lifted from Zephyr to SOF simply did not compile anymore.

ujfalusi replied:
So this will _only_ work with zephyr, no xtos support?

lgirdwood replied:
> So this will _only_ work with zephyr, no xtos support?

Yes. No point doing this for xtos.

ujfalusi commented on src/probe/probe.c:
```diff
@@ -945,20 +948,26 @@ int probe_point_add(uint32_t count, struct probe_point *probe)
-			return -EINVAL;
+		if (probe[i].buffer_id == PROBE_LOGGING_BUFFER_ID_MAGIC) {
+			/* special case for logging buffers */
+			dev = 0;
```
dev = NULL;

ujfalusi commented on src/probe/probe.c:
```diff
@@ -915,6 +917,7 @@ int probe_point_add(uint32_t count, struct probe_point *probe)
 	uint32_t first_free;
 	uint32_t dma_found;
 	struct ipc_comp_dev *dev;
+	struct comp_buffer *caller_id;
```
dev can be local within the else {} path from line 956,

s/caller_id/cb

you don't care about dev anywhere, you only need the cb for the `notifier_register`,

Save the cb at line 953: `cb = logging_comp.cb;` instead of dev
Save the cb after line 970: `cb = dev->cb;`

Use the saved `cb` in the `notifier_register()` calls instead of `dev->cb`

lgirdwood reviewed (commented):
@kv2019i what are the next steps now ?

kv2019i commented:
@lgirdwood wrote:
> @kv2019i what are the next steps now ?

I'm close to complete an update that gets rid of the buffer-id overload, and simplifies the interface between logger and probe code. Thanks @ujfalusi  for the feedback for the interface!

Originally I was thinking that I keep this as draft until @jsarha and @abonislawski have the IPC4 updates ready, and then we merge probes logging as a IPC4 only feature. But, but, the latest versions starts to be so clean, so maybe we start with a IPC3 probes logging imlementation first. This can be used to kick off user-space tool development and testing, and then when the IPC4 work is ready, we can add that on top. I'll push a new PR update today.

kv2019i commented:
V3 updated:
 - much cleaner implementation (direct hook callback, no need to go via comp_buffer, audio_stream and others)
 - small ABI change to add a new probe purpose type
 - open: the probe_dma_buf.avail field make the code a bit clumsy.. with audio probes this was not a problem as the probed data was nicely aligned with DMA minimum aligned transfer size, but with logs, the code becomes a bit messy... can be cleaned up 
 - open: need to figure whether there is a good header/format to use for logging probe data... now setting format with zeros and our SOF probes decoder is happy with this
 - open: reliability issues -- occasional hits a Zephyr panic, needs more testing

ujfalusi commented on src/probe/probe.c:
```diff
@@ -1021,10 +1066,12 @@ int probe_point_add(uint32_t count, struct probe_point *probe)
 			}
-		} else if (probe[i].purpose == PROBE_PURPOSE_EXTRACTION) {
+		} else if (probe[i].purpose == PROBE_PURPOSE_EXTRACTION ||
+			   probe[i].purpose == PROBE_PURPOSE_LOGGING) {
```
Earlier the code have already checked the validity of the purpose (INJECTION, EXTRACTION or LOGGING). Here if the purpose was not INJECTION then it is either EXTRACTION or LOGGING, thus a simple `} else {` is fine?

ujfalusi replied:
Applies to the original code as well, not related to this patch.

kv2019i replied:
Will fix.

ujfalusi commented on src/probe/probe.c:
```diff
@@ -1021,10 +1066,12 @@ int probe_point_add(uint32_t count, struct probe_point *probe)
 				if (_probe->probe_points[j].stream_tag != PROBE_DMA_INVALID &&
-				    _probe->probe_points[j].purpose == PROBE_PURPOSE_EXTRACTION)
+				    (_probe->probe_points[j].purpose == PROBE_PURPOSE_EXTRACTION ||
+				     _probe->probe_points[j].purpose == PROBE_PURPOSE_LOGGING))
```
```c
if (_probe->probe_points[j].stream_tag != PROBE_DMA_INVALID &&
    _probe->probe_points[j].purpose != PROBE_PURPOSE_INJECTION)
```
perhaps? Or can the purpose of the point be other than the valid three at any given time?

kv2019i replied:
I added an explicit check with a helper to make the code more readable.

ujfalusi commented on src/probe/probe.c:
```diff
@@ -1041,10 +1088,14 @@ int probe_point_add(uint32_t count, struct probe_point *probe)
-				  &probe_cb_produce, 0);
-		notifier_register(_probe, dev->cb, NOTIFIER_ID_BUFFER_FREE,
-				  &probe_cb_free, 0);
+		if (probe[i].purpose == PROBE_PURPOSE_EXTRACTION) {
```
Should this be `if (probe[i].purpose != PROBE_PURPOSE_LOGGING)` instead?

kv2019i replied:
Oh yes, buffer-id-produce notifier is also used for injection, will update.

kv2019i commented on src/probe/probe.c:
```diff
@@ -606,7 +606,7 @@ static int copy_from_pbuffer(struct probe_dma_buf *pbuf, void *data,
  * \return 0 on success, error code otherwise.
  */
-static int probe_gen_header(struct comp_buffer *buffer, uint32_t size,
+static int probe_gen_header(uint32_t buffer_id, uint32_t size,
```
Oops, will update.

kv2019i commented on src/probe/probe.c:
```diff
@@ -778,7 +778,7 @@ static void probe_cb_produce(void *arg, enum notify_id type, void *data)
 					  buffer->stream.rate,
 					  buffer->stream.channels);
-		ret = probe_gen_header(buffer,
+		ret = probe_gen_header(buffer->id,
```
Ack @lyakh , there is quite a bit of work, but not really specific to this PR for logging. See also https://github.com/thesofproject/sof/issues/6100

kv2019i commented on src/probe/probe.c:
```diff
@@ -737,6 +761,23 @@ static uint32_t probe_gen_format(uint32_t frame_fmt, uint32_t rate,
+	struct probe_pdata *_probe = probe_get();
+	int ret;
+
+	ret = probe_gen_header(PROBE_LOGGING_BUFFER_ID_MAGIC, length, 0);
```
Did more investigation today. I'll change this to 0 as that is never used for audio buffers. Not perfect, but most pragmatic solution and works with existing tools that process probe output stream.

kv2019i commented on src/probe/probe.c:
```diff
@@ -31,7 +31,7 @@ DECLARE_TR_CTX(pr_tr, SOF_UUID(probe_uuid), LOG_LEVEL_INFO);
 		 0x81, 0x2f, 0xf2, 0xd5, 0xe4, 0xf1, 0x9e, 0x4a);
 
-LOG_MODULE_REGISTER(probe, CONFIG_SOF_LOG_LEVEL);
+LOG_MODULE_REGISTER(probe, LOG_LEVEL_NONE);
```
Aa, sorry, need to add more text here. There' s a problem with recursive calls to logging subsystem. The probe code has logging call itself, which will cause the backend to be called. Not sure how to solve this cleanly yet. This LOG_LEVEL_NONE is the best attempt so far, but we probably need something more robust. I'm sure this is a problem for other logging backends as well.

kv2019i replied:
The V5 set simply allows logging from within the backend and Zephyr logging backend core seems to survive this. So leaving this as it is now.

kv2019i commented on src/include/sof/probe/probe.h:
```diff
@@ -12,6 +12,29 @@
+ *     with ".probe_id.fields.module_id == 0".
+ *   - IPC3 does not have such a field, in IPC3 logging
+ *     prototype, a magic buffer identifier that is known
+ *     both by host and the FW, is used.
```
Ack, removed.

kv2019i commented on src/include/ipc/probe.h:
```diff
@@ -22,6 +22,7 @@
 
 #define PROBE_PURPOSE_EXTRACTION	0x1
 #define PROBE_PURPOSE_INJECTION		0x2
+#define PROBE_PURPOSE_LOGGING		0x3
```
Ack, will update.

kv2019i commented on src/probe/probe.c:
```diff
@@ -200,13 +200,31 @@ static int probe_dma_deinit(struct probe_dma_ext *dma)
 					      (void *)_probe->ext_dma.dmapb.r_ptr,
-					      _probe->ext_dma.dmapb.avail);
+					      avail);
+	}
```
Hmm, maybe yes. With audio this has (apparently?) not been a problem as audio data comes in nicely align blobs of data and the current logic in code has worked. But with logging data (that may come 1 byte at a time), I needed to fix many corner cases in the code. But yeah, could be in a separate patch.

kv2019i replied:
A separate patch now.

kv2019i commented on src/probe/probe.c:
```diff
@@ -216,12 +234,18 @@ static enum task_state probe_task(void *data)
 
+static void probe_logging_hook(uint8_t *buffer, size_t length);
+
+
```
Right, I used to call this from the init(). No longer needed, will remove.

kv2019i commented on src/probe/probe.c:
```diff
@@ -737,6 +761,23 @@ static uint32_t probe_gen_format(uint32_t frame_fmt, uint32_t rate,
+	if (ret < 0)
+		return;
+
+	probe_task(NULL);
```
@lyakh That's a very good question. This is used like this in current code. But yeah, this may not be a safe approach anymore.

kv2019i replied:
@lyakh  changed these into reschedule_task() calls

kv2019i commented on src/trace/log_backend_probe.c:
```diff
@@ -0,0 +1,121 @@
+static struct comp_buffer *buffer;
+
+struct ipc_comp_dev logging_comp;
+static probe_logging_hook_t hook;
```
@lyakh This is following style of upstream Zephyr backends. I can modify to be a bit more SOF style as this will be in SOF git.

kv2019i commented:
V4 update:
 - functionality now ready
 - interface changes ready, needs a slight update to the demux tool -> https://github.com/thesofproject/sof/pull/6130/
 - adds a new Kconfig for the backend (only available in Zephyr builds)
 - fix XTOS SOF build
 - split generic probe_task() fixes to a separate commit
 - todo: cleanup related to "probe.purpose" checks (raised in review)
 - todo: double-check the logic to manage "avail checks" in probe_task()
 - todo: add probe logging to as a default for Intel IPC4 platforms (at least), and disable dma-trace

kv2019i commented:
V5 updated: 
- no longer draft, ready for review
- tested with multiple log formats (text, dictionary, mipi -syst)
- a few major bugfixes that caused the logging to end in FW panic
- places for further cleanup: 
      - the DMA buffer handling code is still a bit messy
      - the configurability via Kconfig is limited compared to upstream Zephyr logging backends
- work needed with "crecord" tool (not related to this PR)
      - logs stay buffered in kernel for a long time
      - drain is apparently not working, so end of log stream is lost due to Linux kernel issue, possible related to https://github.com/thesofproject/linux/pull/3808

lgirdwood commented on src/include/ipc/probe.h:
```diff
@@ -109,7 +110,9 @@ struct probe_dma {
 	uint32_t buffer_id;	/**< ID of buffer to which probe is attached */
-	uint32_t purpose;	/**< PROBE_PURPOSE_EXTRACTION or PROBE_PURPOSE_INJECTION */
+	uint32_t purpose;	/**< PROBE_PURPOSE_EXTRACTION, PROBE_PURPOSE_INJECTION or
+				  *  PROBE_PURPOSE_LOGGING
```
Can you add TRACE here too (we can do in 1 ABI bump) as LOG and TRACE are different things in Zephyr.

kv2019i replied:
@lgirdwood I can add if it's ok to have it without matching implementation in this PR.

lgirdwood reviewed (commented):
LGTM, and I see if picks up the buffer ID change from @lyakh PR too.

lyakh commented on src/probe/probe.c:
```diff
@@ -926,6 +958,14 @@ static void probe_cb_free(void *arg, enum notify_id type, void *data)
+	if (purpose == PROBE_PURPOSE_EXTRACTION || purpose == PROBE_PURPOSE_LOGGING)
+		return true;
+
+	return false;
```
nitpick: `return purpose == PROBE_PURPOSE_EXTRACTION || purpose == PROBE_PURPOSE_LOGGING;`

kv2019i replied:
Change in V6.

lyakh commented on src/trace/log_backend_probe.c:
```diff
@@ -0,0 +1,111 @@
+	k_spin_unlock(&probe_lock, key);
+}
+
+static inline void probe_log_dropped(const struct log_backend *const backend,
```
this cannot be `inline` - its pointer is used

kv2019i replied:
Oops, similar pattern in Zephyr side backends. FYI @teburd  Will fix here.

jsarha commented:
> term1# crecord -c3 -d0 -b8192 -f4 -FS32_LE -R48000 -C4 /tmp/extract.dat
> term2# echo 0,3,0 > /sys/kernel/debug/sof/probe_points
> term1# <terminate crecord>
> term1# sof-probes -p /tmp/extract.dat
> term1# cat buffer_999999.wav

I wonder what I am missing. The echo steps does not produce an error and the probe point is there:
root@jhary:~# cat  /sys/kernel/debug/sof/probe_points
Id: 0x00000000  Purpose: 3  Node id: 0x2

Still there is no data captured by crecord ("We read 8192, DSP sent 0" all the time), and obviously nothing when I try to extract the capture. I have this branch and "CONFIG_LOG_BACKEND_SOF_PROBE=y" in my confing in addition to old probes options (PCM probes work), should I need something more?

kv2019i commented:
V6 uploaded:
- moved the backend code from sof/src/trace/ to sof/src/logging/
- addressed comments by @lyakh  and @lgirdwood
